packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -410,5 +413,10 @@ export class PendingStateManager implements IDisposable {
 				);
 			}
 		}
+
+		this.logger?.sendTelemetryEvent({
```

Comment:
Two ways how this is different from the previous implementation:
- it might record same event twice - once for "read" connection, and once for "write" connection. I'm saying "might" as Wes recently changed logic in Loader layer to leverage "dirty" state of the container to determine connection mode, but I believe it will not change from "read" to "write" if reconnect request is already in-flight and we have some local changes made while that's the case (which I think will be the dominant case: a lot of changes made in offline).
- we were recording zero before

I'd recommend adding clientId on event for easier correlation analyzes.

Reply:
> read-only != "read" connection. DeltaManager.active tracks "read" vs. "write" connections.

I'm not sure what you're referring to re: DeltaManager.active, I don't see anything checking that on this flow?  ContainerRuntime.canSendOps() checks deltaManager.readOnlyInfo.readonly, which should report false if readonly for any reason (including permissions, forced, etc.).  So regardless of why we are "readonly", we won't PendingStateManager.replayPendingStates() and therefore won't log this telemetry, which I think is correct.

> > second call would early-exit in PendingStateManager.replayPendingStates since the pending messages would have already been flushed out by the first call.
> 
> That's not how it works. "flushed" here does not mean - removal of messages form pending state manager - they stay there until we get an ack (well, they go through rebase process, but unless we inbounded some messages, we are getting back exactly same messages). So if we reconnect 5 times and never get a single ack, then we will send same messages 5 times.

If we actually do resubmit the messages 5 times then I think it's appropriate to log this telemetry 5 times - and this is what we already do with current telemetry anyway.

Reply:
read-only storage level permissions (and by extention - readonly mode of a container) is a mode where ops are not flowing to socket at all. In this mode ruintime will never be "connected", as loader layer lies to runtime and leaves it in the dark, as if we are never connected (ops are inbound, but runtime does not differentiate and does not know how they are coming in). This mode is irrelevant for this discussion.

"Read" connection mode vs. "write" connection mode are relevant - runtime can observe the difference through DeltaManager.active. We prefer most of the code not to care though. If container connected as "read", then it will try to submit ops, but they will not go anywhere - this will cause loader layer to drop "read" connection and reconnect as "write". Runtime will resubmit all the ops again.

In context of this change it's useful to look at how we will use this telemetry. We might want to answer a question - how many times user moved out from offline (or temporary offline) with ops or without. In order to answer that question, it's useful to ignore empty attempts of runtime to submit ops in "read" mode connection.

So I'd advice either not to issue this event for "read" connections, or at least add a property to an event that would indicate type of connection (active thing). This will allow much simpler analyzes of data.
Not having ops - I'd rather log it always. But likely either way is fine as we can adjust queries to account for difference in telemetry. Doing math becomes harder (like calculating P95 of ops to resubmit on reconnect - can't use Kusto built-in function directly, need to do math on top of the number that Kusto spits out to account for it).

Reply:
Ok, I think I understand what you're saying now - 
1. DeltaManager.readOnlyInfo.readonly is false not just on write connections but also on read connections.  This is because it's not talking about the current connection but rather something like "could I get a write connection if I wanted to"
2. Therefore ContainerRuntime.canSendOps() can return true when it actually can't send ops (it's on a readonly connection).  So it's really saying something more like "I **could** reconnect in write mode to send the ops".
3. Instead of explicitly disconnecting the readonly connection and reconnecting in write mode before replaying the ops, we trust that attempting to replay the ops will get us nacked and implicitly trigger a reconnection in the write mode we need.
4. This act of attempting to replay the ops on the read connection will send the telemetry added in this PR.  This is because we are in fact replaying the ops (just with no hope of them actually going through).

I feel like there are a number of things that are confusing about that flow that we should probably examine, but for now I'll add an `active` check to filter out the read connections and also remove the early-exit on zero pending messages so we continue to log zero.